### PR TITLE
convert text component to TS

### LINKED ISSUE #18325: Migrate components to TS: Text
### Description
Migrate `Text` component to TypeScript

As apart of the extension TypeScript migration effort and to align with the [Core lane strategy of great to build on](https://docs.google.com/document/d/12J26BKIhMjRpeuDPdBJGkubVIabq8MDo46DhJG1eZzw/edit#heading=h.dnyyhdrgw7bs). All of our current components in the `component-library` should be migrated to TypeScript.

This is high priority and beneficial for a number of reasons:
- Eventually we will have to do this. Sooner than later is better because it will be less complicated now
- It will promote adoption of new components. If you want to create a feature in TS using all DS components will enable you to do so
- Aligns with other teams in the core lane and improves the stability of the extension
- Feature teams are not blocked by the DS creating components we can resume component creation once this has been completed.

### Technical Details
- Convert Objects in `SNAKE_CASE` to enums in `PascalCase` follow [TS conventions](https://www.notion.so/Use-TypeScript-9ec7bd8fc8cf4d82b81cd8943c4707cd?pvs=4#7e6410b21100455a9a08442108f19d78)
- Update file name from `component-name.contants.js` to `component-name.types.ts` and add types
- Update file name from `component-name.js` to `component-name.tsx` and add types
- Update file name from `component-name.test.js` to `component-name.test.tsx`
- Update file name from `component-name.stories.js` to `component-name.stories.tsx`
- Update all instances of component objects to enums
- Update all documentation 

### Acceptance Criteria
- Component has been converted to TypeScript

### PR DESCRIPTION
**Pull request by garrettbear**
## Explanation
* Fixes: #18325 

Update `Text` component to TS

## Manual Testing Steps
`yarn jest ui/components/component-library`

## Pre-merge author checklist

- [X] I've clearly explained:
  - [X] What problem this PR is solving
  - [X] How this problem was solved
  - [X] How reviewers can test my changes
- [X] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [X] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [X] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/components/component-library/text/text.stories.tsx [hunk 1]
```diff
@@ -16,115 +17,49 @@ import {
...
-    color: {
-      control: { type: 'select' },
-      options: Object.values(TextColor),
-    },
+    // variant: {
+    //   options: Object.values(TextVariant),
+    // },
+    // color: {
+    //   options: Object.values(TextColor),
+    // },
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.stories.tsx:39]: Should these be uncommented or removed?

File: ui/components/component-library/text/text.stories.tsx [hunk 2]
```diff
@@ -16,115 +17,49 @@ import {
...
-      table: { category: 'box props' },
-    },
-    marginLeft: {
-      options: marginSizeKnobOptions,
-      control: 'select',
-      table: { category: 'box props' },
-    },
+    // textDirection: {
+    //   options: TextDirection,
+    // },
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.stories.tsx:60]: Should this be uncommented or removed?

File: ui/components/component-library/text/text.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    case TextVariant.displayMd:
+      return 'h1';
+    case TextVariant.headingLg:
+      return 'h2';
+    case TextVariant.headingMd:
+      return 'h3';
+    case TextVariant.headingSm:
+      return 'h4';
+    case TextVariant.inherit:
+      return 'span';
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.tsx:22]: Should these return the ValidTag enum?

File: ui/components/component-library/text/text.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    textAlign,
+    textDirection,
+    overflowWrap,
+    ellipsis,
+    as,
+    className = '',
+    children,
+    ...props
+  }: TextProps,
+  ref: Ref<HTMLDivElement>,
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.tsx:45]: Should this be any HTML element? `h1-h6` `p`, `li` etc
```suggestion
  ref: Ref<HTMLElement>,
```

File: ui/components/component-library/text/text.types.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  /**
+   * The color of the Text component Should use the COLOR object from
+   * ./ui/helpers/constants/design-system.js
+   */
+  color?: TextColor | Color;
+  /**
+   * The font-weight of the Text component. Should use the FONT_WEIGHT object from
+   * ./ui/helpers/constants/design-system.js
+   */
+  fontWeight?: typeof FONT_WEIGHT;
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.types.ts:76]: Should we use keyof here? We don't want the whole object just the key value.
```suggestion
  fontWeight?: keyof typeof FONT_WEIGHT;
```

File: ui/components/component-library/text/text.types.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
+import React from 'react';
+import { BoxProps } from '../../ui/box';
```
Comment (georgewrmarshall) [inline ui/components/component-library/text/text.types.ts:2]: I think you want
```suggestion
import type { BoxProps } from '../../ui/box/box.d';
```

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
Good start. Left some suggestions

**Review by Jah2233 — COMMENTED** [review]
Слава Україні!

**Comment by Jah2233** [comment]
Дякую

**Review by georgewrmarshall — COMMENTED** [review]
Do you need to create the deprecated folder if you aren't updating the types to enums?

**Review by georgewrmarshall — COMMENTED** [review]
Left one suggestion. I pulled the branch and it seems to work?

**Review by georgewrmarshall — APPROVED** [review]

**Review by NidhiKJha — APPROVED** [review]

**Pull request merged**